tests/macros/src/main/scala/scala/macros/tests/sourcecode/SourceContext.scala

```diff
@@ -0,0 +1,77 @@
+package scala.macros.tests.sourcecode
+
+import scala.macros._
+import scala.language.implicitConversions
+
+// Unification of macro impl/defn forces us to put utilities in separate object
+private[sourcecode] object Util {
+  // TODO(olafur) replace with more typesafe typeOf[Line]
+  private def root: Term =
+    Term.Name("_root_").select("scala" :: "macros" :: "tests" :: "sourcecode" :: Nil)
+  def prefix(name: String): Term =
+    root.select(name :: Nil)
+
+  // Should this be handled by engines? I would claim that this is application specific
+  // for sourcecode.
+  def isSynthetic(s: Symbol)(implicit m: Mirror): Boolean = {
+    val name = s.name.value
+    name == "<init>" ||
+    name == "<root>" ||
+    (name.startsWith("<local ") && name.endsWith(">"))
+  }
+
+  def symbolName(s: Symbol)(implicit m: Mirror): String = s.name.value.stripSuffix("$")
+}
+
+case class File(value: String)
+object File {
+  implicit def generate: File = macro {
+    val path = enclosingPosition.input.path.toAbsolutePath.toString
+    Util.prefix("File").apply(Lit.String(path) :: Nil)
+  }
+}
+
+case class Line(value: Int)
+object Line {
+  implicit def generate: Line = macro {
+    Util.prefix("Line").apply(Lit.Int(enclosingPosition.line) :: Nil)
+  }
+}
+
+case class Name(value: String)
+object Name {
+  implicit def generate: Name = macro {
+    def loop(s: Symbol): Symbol =
+      if (Util.isSynthetic(s)) s.owner.fold(s)(loop)
+      else s
+    val owner = loop(enclosingOwner)
+    Util.prefix("Name").apply(Lit.String(Util.symbolName(owner)) :: Nil)
+  }
+}
+
+case class FullName(value: String)
+object FullName {
+  implicit def generate: FullName = macro {
+    def loop(sym: Symbol): Vector[Symbol] = {
+      val owner = sym.owner.fold(Vector.empty[Symbol])(loop)
+      if (!Util.isSynthetic(sym)) {
+        owner :+ sym
+      } else owner
+    }
+    val names = loop(enclosingOwner)
+    val fullName = names.map(Util.symbolName).mkString(".")
+    Util.prefix("FullName").apply(Lit.String(fullName) :: Nil)
+  }
+}
+
+case class Text[T](source: String, value: T)
+object Text {
+  implicit def generate[T](e: T): Text[T] = macro {
+    // NOTE(olafur): lihaoyi/sourcecode uses unit parser and slices
+    // start/end offsets from input.content:
+    // https://github.com/lihaoyi/sourcecode/blob/420bea7941d3219e2f1200b14b11010843aea39c/sourcecode/shared/src/main/scala/sourcecode/SourceContext.scala#L137-L140
+    // This produces more accurate/portable results.
+    val source = e.syntax
```

Comment:
This may not always work currently, e.g.  when `e` contains closure. Going into source code is also unsafe, as the trees might be synthesized, not in the source code.

Reply:
That's fine, this value is only used for debugging. The actual sourcecode macro slices from `Input.content`, but we don't have support for range positions #28